### PR TITLE
[SPARK-18141][SQL] Fix to quote column names in the predicate clause  of the JDBC RDD generated sql statement

### PR DESCRIPTION
**Pull request by sureshthalamati**
## What changes were proposed in this pull request?

SQL query generated for the JDBC data source is not quoting columns in the predicate clause. When the source table has quoted column names,  spark jdbc read fails with column not found error incorrectly.  

Error:
org.h2.jdbc.JdbcSQLException: Column "ID" not found; 
Source SQL statement:
SELECT "Name","Id" FROM TEST."mixedCaseCols" WHERE (Id < 1)

This PR fixes by quoting column names in the generated  SQL for predicate clause  when filters are pushed down to the data source.

Source SQL statement after the fix:
SELECT "Name","Id" FROM TEST."mixedCaseCols" WHERE ("Id" < 1)

## How was this patch tested?

Added new test case to the JdbcSuite

### CODE REVIEW

**Comment by sureshthalamati** [comment]
Test failed is org.apache.spark.sql.streaming.StreamingQuerySuite, unrelated to this change. Might have been fixed in commit https://github.com/apache/spark/commit/79fd0cc0584e48fb021c4237877b15abbffb319a

**Comment by sureshthalamati** [comment]
retest this please

**Comment by sureshthalamati** [comment]
@rxin @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -105,35 +105,38 @@ object JDBCRDD extends Logging {
    * Turns a single Filter into a String representing a SQL expression.
    * Returns None for an unhandled filter.
    */
-  def compileFilter(f: Filter): Option[String] = {
+  def compileFilter(f: Filter, dialect: JdbcDialect): Option[String] = {
     Option(f match {
-      case EqualTo(attr, value) => s"$attr = ${compileValue(value)}"
+      case EqualTo(attr, value) => s"${dialect.quoteIdentifier(attr)} = ${compileValue(value)}"
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Add a nested function in compileFilter 
```Scala
def quote(colName: String): String = dialect.quoteIdentifier(colName)
```

Then, your code changes can look cleaner.

**Comment by sureshthalamati** [comment]
Thank you very much for the feed back @gatorsmile . Addressed the review comments.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -790,4 +807,23 @@ class JDBCSuite extends SparkFunSuite
...
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id < 1").collect().size == 0)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id <= 1").collect().size == 1)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id > 1").collect().size == 2)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id >= 1").collect().size == 3)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id = 1").collect().size == 1)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id != 2").collect().size == 2)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Id <=> 2").collect().size == 1)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE Name LIKE 'fr%'").collect().size == 1)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE NAME LIKE '%ed'").collect().size == 1)
+    assert(sql("SELECT * FROM mixedCaseCols WHERE NAME LIKE '%re%'").collect().size == 1)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: In the above two statements, `NAME ` -> `Name`

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: What is the purpose of the above two statements?

**Comment by gatorsmile** [comment]
This sounds a right and critical fix to me; otherwise we are unable to resolve the columns of predicates in the case sensitive JDBC sources. 

@sureshthalamati Could you post the following exception in your PR description? 
```
org.h2.jdbc.JdbcSQLException: Column "ID" not found; SQL statement:
SELECT "Name","Id" FROM TEST."mixedCaseCols" WHERE (Id < 1) [42122-183]
```

 cc @srowen Could you please check it? Any comment? Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 2 above)
Comment (sureshthalamati) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Those two statements test  String StartsWith , and Contains filters. They are pushed to jdbc data source, and mapped to SQL LIKE expression.

I will fix the inconsistent column name in above two statements.

**Comment by sureshthalamati** [comment]
Thank you for reviewing, @gatorsmile .  Updated the PR description , and addressed all the review comments.

**Comment by sureshthalamati** [comment]
@gatorsmile I addressed all the review comments , can you please take a look.

**Comment by gatorsmile** [comment]
@srowen Any comment on this?

**Comment by gatorsmile** [comment]
@sureshthalamati Could you resolve the conflict? Thanks!

**Comment by sureshthalamati** [comment]
Thanks, @gatorsmile . Resolved the conflicts, and also added test case for empty in clause with mixed case column name.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 3]
```diff
@@ -855,6 +855,8 @@ class JDBCSuite extends SparkFunSuite
     assert(sql("SELECT * FROM mixedCaseCols WHERE Name LIKE '%re%'").collect().size == 1)
     assert(sql("SELECT * FROM mixedCaseCols WHERE Name IS NULL").collect().size == 1)
     assert(sql("SELECT * FROM mixedCaseCols WHERE Name IS NOT NULL").collect().size == 2)
+    assert(sql("SELECT * FROM mixedCaseCols")
+      .filter($"Name".isin(Array[String]() : _*)).collect().size == 0)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]:
```
.filter($"Name".isin(Array[String]() : _*)).collect().size == 0)
```
->
```Scala
.filter($"Name".isin()).collect().size == 0)
```

**Comment by gatorsmile** [comment]
LGTM except a minor comment 

cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 3 above)
Comment (sureshthalamati) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Thanks , @gatorsmile . Fixed it.

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
Merging to master/2.1! Thanks!

**Comment by sureshthalamati** [comment]
Thank you , @gatorsmile @cloud-fan

**Pull request closed without merging**